youtube/yt_data_extract/watch_extraction.py

```diff
@@ -936,13 +966,154 @@ def decrypt_signatures(info):
+                n_sig_result = decrypt_n_signature(n_sig=n_sig, jscode=js_nsig_decrypt_code)
+                print('Replacing n with n_sig_result')
+                print('Current n value: ' + params[n_sig_index])
+                params.pop(n_sig_index)
+                params.insert(n_sig_index, 'n=' + n_sig_result)
+                print('n value now is: ' + params[n_sig_index])
+                final_url = '&'.join(params)
+                fmt['url'] = final_url
     return False
+
+# nsig decryption
+# adapted from iv-org/inv-sig-helper
+# regex matching will be performed using flpc (imported as regex) instead of python's built in re for performance reason.
+NSIG_FUNCTION_ARRAYS = [
+    r'null\)&&\([a-zA-Z]=(?P<nfunc>[a-zA-Z0-9$]+)\[(?P<idx>\d+)\]\([a-zA-Z0-9]\)',
+    r'(?x)&&\(b="n+"\[[a-zA-Z0-9.+$]+\],c=a\.get\(b\)\)&&\(c=(?P<nfunc>[a-zA-Z0-9$]+)(?:\[(?P<idx>\d+)\])?\([a-zA-Z0-9]\)',
+]
+
+NSIG_FUNCTION_ENDINGS = [
+    r'=\s*function(\([\w]+\)\{\s*var\s+[\w\s]+=[\w\.\s]+?\.call\s*\([\w\s$]+?,[\(\)\",\s]+\)[\S\s]*?\}\s*return [\w\.\s$]+?\.call\s*\([\w\s$]+?\s*,[\(\)\",\s]+\)\s*\}\s*;)',
+    r'=\s*function([\S\s]*?\}\s*return \w+?\.join\(\"\"\)\s*\};)',
+    r'=\s*function([\S\s]*?\}\s*return [\W\w$]+?\.call\([\w$]+?,\"\"\)\s*\};)',
+]
+
+def extract_nsig_func(base_js):
+    for i, member in enumerate(NSIG_FUNCTION_ARRAYS):
+        func_array_re = regex.compile(member.replace('$', '\\$'))
+        func_array = regex.search(func_array_re, base_js)
+        if not func_array == None:
+            func_name = func_array.group(1)
+            break
+    NSIG_CONTEXT = 'var '+func_name+'\\s*=\\s*\\[(.+?)][;,]'
+    func_context = regex.search(regex.compile(NSIG_CONTEXT), base_js)
+    func_body_re = []
+    for i, member in enumerate(NSIG_FUNCTION_ENDINGS):
+        func_body_re_item = ''
+        func_body_re_item += func_context.group(1)
+        func_body_re_item += member
+        func_body_re.append(func_body_re_item)
+    nsig_func_body = None
+    for i, member in enumerate(func_body_re):
+        result = regex.search(regex.compile(member.replace('$', '\\$')), base_js)
+        if result != None:
+            nsig_func_body = result.group(1)
+            print('nsig_func_body length: ' + str(len(nsig_func_body)))
+            break
+    if nsig_func_body != None:
+        nsig_func_name = 'decrypt_nsig'
+        full_nsig_func = 'var ' + nsig_func_name + '=function' + nsig_func_body
+        return full_nsig_func
+    else:
+        return None
+
+def decrypt_n_signature(n_sig, jscode):
+    print('jscode len is: ' + str(len(jscode)))
+    dukpy_session = dukpy.JSInterpreter()
+    # Loading the function into dukpy session
+    dukpy_session.evaljs(jscode)
+    print('n_sig = ' + n_sig)
+    #n_sig_result = dukpy_session.evaljs('decrypt_nsig("' + n_sig + '")')
+    n_sig_result = dukpy_session.evaljs("decrypt_nsig(dukpy['n'])", n=n_sig)
+    print('n_sig_result = ' + n_sig_result)
+    return n_sig_result
```

Comment:
`dukpy` is just a python wrapper for [duktape](https://github.com/svaarala/duktape) js engine.

The pypi package of `dukpy` has `dukpy-install` command which is able to download npm packages from the internet.

Unless told to do so, `dukpy` module doesn't access the internet for as far as I know. The `nsig_func` doesn't need access to the internet during runtime, which I have verified doing manual `n_sig` decryption using various python js bindings.

I also consider `dukpy` as just-work lightweight js engine for python, since it has wheels for `arm64` on pypi and `armhf` on piwheels.org so if anyone runs this on their single board computers, they will hopefully meet no problems during runtime.